// svelteplot/dist/transforms/recordize.js
import isDataRecord from '../helpers/isDataRecord.js';
import { INDEX } from '../constants';
export const X = Symbol('x');
export const Y = Symbol('y');
export const RAW_VALUE = Symbol('originalValue');
export function indexData(data) {
    return data.map((d, i) => ({ ...d, [INDEX]: i }));
}
/*
 * This transform takes an array of raw values as input and returns data records
 * in which the values are interpreted as x channel and their index as y
 */
export function recordizeX({ data, ...channels }, { withIndex } = { withIndex: true }) {
    const dataIsRawValueArray = !isDataRecord(data[0]) && !Array.isArray(data[0]) && channels.x == null;
    if (dataIsRawValueArray) {
        // we remove x, x1 and x2 from the channels since they make no sense when
        // the data is a raw value array
        const { x, x1, x2, ...nonXChannels } = channels;
        return {
            data: data.map((value, index) => ({
                [RAW_VALUE]: value,
                [INDEX]: index
            })),
            ...nonXChannels,
            x: RAW_VALUE,
            ...(withIndex ? { y: INDEX } : {})
        };
    }
    return { data: indexData(data), ...channels };
}
/*
 * This transform takes an array of raw values as input and returns data records
 * in which the values are interpreted as y channel and their index as yx
 */
export function recordizeY({ data, ...channels }, { withIndex } = { withIndex: true }) {
    if (!data)
        return { data, ...channels };
    const dataIsRawValueArray = !isDataRecord(data[0]) && !Array.isArray(data[0]) && channels.y == null;
    if (dataIsRawValueArray) {
        // we remove y, y1 and y2 from the channels since they make no sense when
        // the data is a raw value array
        const { y, y1, y2, ...nonYChannels } = channels;
        return {
            data: Array.from(data).map((value, index) => ({
                [INDEX]: index,
                [RAW_VALUE]: value
            })),
            ...nonYChannels,
            ...(withIndex ? { x: INDEX } : {}),
            y: RAW_VALUE
        };
    }
    return { data: indexData(data), ...channels };
}
/**
 * This transform is used to allow users to pass an [[x0, y0], [x1, y1], ...] array
 * as dataset to marks that support it. It transforms the arrays into records, so
 * the rest of our code doesn't have to deal with this case anymore.
 */
export function recordizeXY({ data, ...channels }) {
    if (!data)
        return { data, ...channels };
    if (!isDataRecord(data[0]) &&
        Array.isArray(data[0]) &&
        channels.x === undefined &&
        channels.y === undefined) {
        return {
            data: data.map(([x, y, ...rest], i) => ({
                [RAW_VALUE]: [x, y, ...rest],
                [INDEX]: i,
                [X]: x,
                [Y]: y
            })),
            ...channels,
            x: X,
            y: Y
        };
    }
    return { data: data, ...channels };
}
export function recordize({ data, ...channels }) {
    if (!data)
        return { data, ...channels };
    if (!isDataRecord(data[0])) {
        return {
            data: data.map((d, i) => ({
                [RAW_VALUE]: d,
                [INDEX]: i
            })),
            ...channels
        };
    }
    return { data: indexData(data), ...channels };
}
