// svelteplot/dist/helpers/getBaseStyles.js
import { resolveProp } from './resolve.js';
/**
 * all style props that can be applied via channels but
 * are not scaled
 */
const styleProps = {
    strokeWidth: 'stroke-width',
    strokeDasharray: 'stroke-dasharray',
    strokeLinejoin: 'stroke-linejoin',
    strokeLinecap: 'stroke-linecap',
    blend: 'mix-blend-mode',
    clipPath: 'clip-path',
    mask: 'mask',
    fontFamily: 'font-family',
    fontSize: 'font-size',
    fontStyle: 'font-style',
    fontWeight: 'font-weight',
    textAnchor: 'text-anchor',
    fontVariant: 'font-variant',
    letterSpacing: 'letter-spacing',
    textDecoration: 'text-decoration',
    textTransform: 'text-transform',
    wordSpacing: 'word-spacing',
    cursor: 'cursor',
    pointerEvents: 'pointer-events'
};
const styleDefaults = {
    fontWeight: 'normal'
};
export function getBaseStylesObject(datum, props) {
    return Object.fromEntries(Object.entries(styleProps)
        .filter(([key, cssKey]) => cssKey && props[key] != null)
        .map(([key, cssKey]) => [
        cssKey,
        maybeToPixel(cssKey, (resolveProp(props[key], datum, styleDefaults[key] || null) ??
            ''))
    ]));
}
export default function (datum, props) {
    return Object.entries(getBaseStylesObject(datum, props))
        .map(([key, value]) => `${key}: ${value}`)
        .join(';');
}
export function maybeToPixel(cssKey, value) {
    if (cssKey === 'font-size' || cssKey === 'stroke-width') {
        return typeof value === 'number' ? `${value}px` : value;
    }
    return value;
}
export function maybeFromPixel(value) {
    return typeof value === 'string' && value.endsWith('px') ? +value.slice(0, -2) : value;
}
export function maybeFromRem(value, rootFontSize = 16) {
    return typeof value === 'string' && value.endsWith('rem')
        ? +value.slice(0, -3) * rootFontSize
        : value;
}
let nextClipId = 0;
let nextPatternId = 0;
export function getClipId() {
    return `svp-clip-${++nextClipId}`;
}
export function getPatternId() {
    return `svp-pattern-${++nextPatternId}`;
}
