// svelteplot/dist/hooks/usePlot.svelte.js
import { getContext } from 'svelte';
/**
 * Internal state representation of a Plot.
 */
class PlotState {
    // Define properties and methods for PlotState as needed
    width = $state(50);
    height = $state(50);
    options = $state({});
    facetWidth = $state(0);
    facetHeight = $state(0);
    plotWidth = $state(0);
    plotHeight = $state(0);
    scales = $state();
    body = $state();
    /**
     * True if there's a color scale and a symbol scale and both are bound to the same
     * single channel accessor.
     */
    colorSymbolRedundant = $state(false);
    /**
     * True if the plot is using filled dot marks.
     */
    hasFilledDotMarks = $state(false);
    css = $state(null);
    constructor(state) {
        // Initialization code here
        Object.assign(this, state);
    }
    update(newState) {
        Object.assign(this, newState);
    }
    get publicState() {
        return new PublicPlotState(this);
    }
}
/**
 * A public-facing wrapper around PlotState that exposes only read-only properties.
 */
class PublicPlotState {
    #plotState;
    constructor(plotState) {
        this.#plotState = plotState;
    }
    get width() {
        return this.#plotState.width;
    }
    get height() {
        return this.#plotState.height;
    }
    get options() {
        return this.#plotState.options;
    }
    get scales() {
        return this.#plotState.scales;
    }
    get plotWidth() {
        return this.#plotState.plotWidth;
    }
    get plotHeight() {
        return this.#plotState.plotHeight;
    }
    get facetWidth() {
        return this.#plotState.facetWidth;
    }
    get facetHeight() {
        return this.#plotState.facetHeight;
    }
    get body() {
        return this.#plotState.body;
    }
    get colorSymbolRedundant() {
        return this.#plotState.colorSymbolRedundant;
    }
    get hasFilledDotMarks() {
        return this.#plotState.hasFilledDotMarks;
    }
    get css() {
        return this.#plotState.css;
    }
}
export function setPlot(initialState) {
    return new PlotState({
        ...initialState
    });
}
export function usePlot() {
    const { getPlotState } = getContext('svelteplot');
    return getPlotState().publicState;
}
